File: prepare_for_interview/question/pascal_triangle.py
from typing import List
def generate_pascal_triangle(depth:int) -> List[List[int]]:
    data = [[1] * (i+1) for i in range(depth)]

    for line in range(2,depth):
        for i in range(1,line):
            data[line][i] = data[line-1][i-1]+data[line-1][i]

    return data

def print_pascal(data:List[int])->None:
    max_digit = len(str(max(data[-1])))
    width = max_digit + (max_digit%2) + 2
    for index,line in enumerate(data):
        numbers = ''.join([str(i).center(width,'*') for i in line])
        print((' '*int(width/2)) * (len(data)-index),numbers)


if __name__ == "__main__":
    
    r = generate_pascal_triangle(10)
    print_pascal(r)



# def get_pascal_triangle(height:int):
#     result = [[] for _  in range(height)]

#     w = 2*height-1
#     mid = w//2

    
    
#     for h in range(height):
#         result[h] = [" " for _ in range(2*height-1)]
#         print(result[0][1])
        
#         if h == 0:
#             result[h][mid] = 1
        
#         if h %2 == 0: 
#             for i in range(h//2+1) :
#                 if h >= 1 and mid-2*i-1 >=0 and mid-2*i+1 < w and mid+2*i-1>=0 and mid+2*i+1 < w:
#                     if result[h-1][mid-2*i-1] != ' ' and result[h-1][mid-2*i+1] != ' ' and result[h-1][mid+2*i-1] != '' and result[h-1][mid+2*i+1] != ' ':
#                         result[h][mid-2*i] = result[h-1][mid-2*i-1] + result[h-1][mid-2*i+1]
#                         result[h][mid+2*i] = result[h-1][mid+2*i-1] + result[h-1][mid+2*i+1]
#                     else:
#                         result[h][mid-2*i] = 1
#                         result[h][mid+2*i] = 1 
#                 else:
#                     result[h][mid-2*i] = 1
#                     result[h][mid+2*i] = 1    
#         else:
#              for i in range(h//2+1) :
#                 if h >= 1 and mid-1-2*i-1 >=0 and mid-1-2*i+1 < w and mid+1+2*i-1>=0 and mid+1+2*i+1 < w:
#                     if result[h-1][mid-1-2*i-1] != ' ' and result[h-1][mid-1-2*i+1] != ' ' and result[h-1][mid+1+2*i-1] != ' ' and result[h-1][mid+1+2*i+1] != ' ':
#                         result[h][mid-1-2*i] = result[h-1][mid-1-2*i-1] + result[h-1][mid-1-2*i+1]
#                         result[h][mid+1+2*i] = result[h-1][mid+1+2*i-1] + result[h-1][mid+1+2*i+1]
#                     else:
#                         result[h][mid-1-2*i] = 1
#                         result[h][mid+1+2*i] = 1 
#                 else:
#                     result[h][mid-1-2*i] = 1
#                     result[h][mid+1+2*i] = 1 
#     return result



# if __name__ == "__main__":
#     for r in get_pascal_triangle(7):
#         print(''.join([str(c) for c in r]))


    

File: prepare_for_interview/question/pascal_triangle.cpp
//                            **1***
//                         **1*****1***
//                      **1*****2*****1***
//                   **1*****3*****3*****1***
//                **1*****4*****6*****4*****1***
//             **1*****5*****10****10****5*****1***
//          **1*****6*****15****20****15****6*****1***
//       **1*****7*****21****35****35****21****7*****1***
//    **1*****8*****28****56****70****56****28****8*****1***
// **1*****9*****36****84***126***126****84****36****9*****1***

// 上図のようなパスカルの三角形を生成しろ

#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> generate_pascal_triangle(int depth)
{

    vector<vector<int>> triangle(depth);

    for (int i = 0; i < depth; ++i)
    {
        triangle[i] = vector<int>(i + 1, 1);
    }

    for (int line = 2; line < depth; ++line)
    {
        for (int i = 1; i < line; ++i)
        {
            triangle[line][i] = triangle[line - 1][i] + triangle[line - 1][i - 1];
        }
    }

    return triangle;
}

void print_pascal(vector<vector<int>> &triangle)
{
    int max_digit = to_string(*max_element(triangle.back().begin(), triangle.back().end())).size();

    int width = max_digit + (max_digit % 2) + 2;

    for (int i = 0; i < triangle.size(); i++)
    {
        cout << string((width / 2) * (triangle.size() - i), ' ');

        for (int num : triangle[i])
            cout << setw(width) << setfill(' ') << num;

        cout << endl;
    }
}

int main()
{

    auto triangle = generate_pascal_triangle(10);
    print_pascal(triangle);

    for (auto line : triangle)
    {
        for (auto e : line)
        {
            cout << e << ",";
        }
        cout << endl;
    }

    return 0;
}

// #include <bits/stdc++.h>
// using namespace std;

// vector<vector<int>> create_triangle(int h)
// {
//     vector<vector<int>> result(h);

//     result[0] = {1};

//     for (int i = 1; i < h; i++)
//     {
//         for (int j = 0; j < i + 1; j++)
//         {
//             result[i].push_back(0);
//         }
//     }

//     for (int i = 1; i < h; i++)
//     {
//         for (int j = 0; j < i + 1; j++)
//         {
//             if (0 <= j - 1 && j < result[i - 1].size())
//                 result[i][j] = result[i - 1][j] + result[i - 1][j - 1];
//             else
//                 result[i][j] = 1;
//         }
//     }

//     return result;
// }

// // パスカルの三角形を印刷する関数
// void print_pascal(const vector<vector<int>> &data)
// {
//     int max_digit = to_string(*max_element(data.back().begin(), data.back().end())).length();
//     int width = max_digit + (max_digit % 2) + 2;

//     for (size_t index = 0; index < data.size(); ++index)
//     {
//         cout << string((width / 2) * (data.size() - index), 't');

//         for (int num : data[index])
//         {
//             cout << setw(width) << setfill('*') << num;
//         }
//         cout << endl;
//     }
// }

// int main()
// {
//     auto result = create_triangle(20);

//     print_pascal(result);

//     return 0;
// }